Adds more descriptive error message to registry

Kleptobismol opened the pull request:
Fixes #1326

ebolyen commented on CHANGELOG.md:
```diff
@@ -5,6 +5,8 @@
 
 ### Features
+* Added more descriptive error message to `skbio.io.registry` when attempting to read without specifying `into` and when there is
+no generator reader. ([#1326](https://github.com/biocore/scikit-bio/issues/1326))
```
single line

ebolyen commented on skbio/io/registry.py:
```diff
@@ -558,11 +558,28 @@ def _init_reader(self, file, fmt, into, verify, kwargs, io_kwargs):
+        for lookup in self._lookups:
+            for format in lookup.values():
+                if format.name == fmt:
+                    return list(format.readers.keys())
```
This can be simplified to 
```python
for lookup in self._lookups:
    if fmt in lookupt:
        return list(lookup[fmt].readers)
```

Kleptobismol replied:
fixed

ebolyen commented on skbio/io/registry.py:
```diff
@@ -558,11 +558,28 @@ def _init_reader(self, file, fmt, into, verify, kwargs, io_kwargs):
         if reader is None:
+            possible_readers = self._get_possible_readers(fmt)
+            if possible_readers:
+                possible_intos = []
```
```python
possible_intos = [r.__name__ for r in self._get_possible_readers(fmt)]
if possible_intos:
    <set message>
else:
    message = ''
```

Kleptobismol replied:
fixed

ebolyen commented on skbio/io/registry.py:
```diff
@@ -558,11 +558,28 @@ def _init_reader(self, file, fmt, into, verify, kwargs, io_kwargs):
+                message = """Possible values for `into` include: %s
+                          """ % ', '.join(possible_intos)
+            else:
+                message = ''
```
Add a test to cover this branch of the code

ebolyen commented on skbio/io/registry.py:
```diff
@@ -558,11 +558,28 @@ def _init_reader(self, file, fmt, into, verify, kwargs, io_kwargs):
-                (fmt, file, into.__name__ if into else 'generator'))
+                "Cannot read %r from %r, no %s reader found. `into` also"
+                " not found. %s" %
+                (fmt, file, into.__name__ if into else 'generator',
```
This ternary suggests that sometimes `into` is set, so the error will be incorrect sometimes.

Kleptobismol replied:
fixed

ebolyen commented:
@Kleptobismol sorry for the delayed review, done now. Have some comments that need to be addressed.

Kleptobismol commented:
@ebolyen addressed your comments, except for the test one, as I'm not sure what you meant. I ran coverage and all branches of the new code are being tested. Also, I had to tweak one of your changes a bit.

ebolyen commented on skbio/io/registry.py:
```diff
@@ -558,11 +558,25 @@ def _init_reader(self, file, fmt, into, verify, kwargs, io_kwargs):
         if reader is None:
+            readers = self._get_possible_readers(fmt)
+            possible_intos = [r.__name__ for r in readers] if readers else None
+            if possible_intos:
```
Oh I see, let's make it so that `_get_possible_readers` always returns a list (empty if there isn't anything). That way we can just chain it with the list comprehension. Then `possible_intos` will be empty (false) or it will have the available class names.

Kleptobismol replied:
fixed

ebolyen commented on skbio/io/registry.py:
```diff
@@ -558,11 +558,25 @@ def _init_reader(self, file, fmt, into, verify, kwargs, io_kwargs):
+            possible_intos = [r.__name__ for r in readers] if readers else None
+            if possible_intos:
+                message = """Possible values for `into` include: %s
+                          """ % ', '.join(possible_intos)
```
Not a huge fan of using triple-quotes here. Could you change to 
```python
message = ("possible values for `into include: %s" 
           % ', '.join(possible_intos))
```
That would match most of our code-base a little better.

Kleptobismol replied:
fixed

ebolyen commented on skbio/io/registry.py:
```diff
@@ -558,11 +558,25 @@ def _init_reader(self, file, fmt, into, verify, kwargs, io_kwargs):
+                          """ % ', '.join(possible_intos)
+            else:
+                message = ''
+            into_message = '`into` also not found.' if not into else ''
```
found -> provided 

might be a better wording

Kleptobismol replied:
fixed

ebolyen commented:
RE: coverage, I think that makes sense.

ebolyen commented:
Looks good @Kleptobismol! Just squash your commits and it should be good for merge.
